### PR TITLE
feat(connector): [BANKOFAMERICA] Implement Google Pay

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Implement Payment Method Google Pay for connector Bankofamerica

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch-cloud/issues/3297

## How did you test it?

Authorize(Manual+Automatic), Capture and PSYNC flow needs to be tested for Payment Method Google Pay for connector Bankofamerica

```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
    
    
    "amount": 900,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    
    "customer_id": "CustomerX",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "wallet",
    "payment_method_type": "google_pay",
    "payment_method_data": {
        "wallet": {
            "google_pay": {
                "type": "CARD",
                "description": "Visa •••• 1111",
                "info": {
                    "card_network": "VISA",
                    "card_details": "1111"
                },
                "tokenization_data": {
                    "type": "PAYMENT_GATEWAY",
                    "token": "GOOGLE_PAY_TOKEN_HERE"
                }
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "Dubai",
            "zip": "94122",
            "country": "AE",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+97"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "AE",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "business_label": "food",
    "business_country": "US"
}'
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 1]
```diff
@@ -97,8 +99,10 @@ pub struct CaptureOptions {
 }
 
 #[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
 pub struct PaymentInformation {
-    card: Card,
+    card: Option<Card>,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:104]: Create enum for PaymentInformation, one for card and one for fluid data

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 2]
```diff
@@ -183,6 +193,122 @@ pub struct ClientReferenceInformation {
...
+        value: (
+            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+            payments::Card,
+        ),
+    ) -> Result<Self, Self::Error> {
+        let (item, ccard) = value;
+        let email = item.router_data.request.get_email()?;
+        let bill_to = build_bill_to(item.router_data.get_billing()?, email)?;
+
+        let order_information = OrderInformationWithBill {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:213]: Handle this with From impl so redundant code can be reduced

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 3]
```diff
@@ -183,6 +193,122 @@ pub struct ClientReferenceInformation {
...
+                number: ccard.card_number,
+                expiration_month: ccard.card_exp_month,
+                expiration_year: ccard.card_exp_year,
+                security_code: ccard.card_cvc,
+                card_type,
+            }),
+            fluid_data: None,
+        };
+
+        let processing_information = ProcessingInformation {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:236]: Use From implementation

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 4]
```diff
@@ -183,6 +193,122 @@ pub struct ClientReferenceInformation {
...
+            capture: matches!(
+                item.router_data.request.capture_method,
+                Some(enums::CaptureMethod::Automatic) | None
+            ),
+            payment_solution: None,
+        };
+
+        let client_reference_information = ClientReferenceInformation {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:244]: Please use From impl

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 3 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 4 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 5]
```diff
@@ -177,12 +199,155 @@ impl From<CardIssuer> for String {
...
+
+        let payment_information = PaymentInformation::GooglePay(GooglePayPaymentInformation {
+            fluid_data: FluidData {
+                value: Secret::from(
+                    consts::BASE64_ENGINE.encode(google_pay_data.tokenization_data.token),
+                ),
+            },
+        });
+
+        let payment_solution = Some("012".to_string());
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:338]: Let's have this as enum, which will be useful while implementing applepay

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 6]
```diff
@@ -177,12 +199,155 @@ impl From<CardIssuer> for String {
...
+        &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+        payments::GooglePayWalletData,
+    )> for BankOfAmericaPaymentsRequest
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        value: (
+            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+            payments::GooglePayWalletData,
+        ),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:322]:
```suggestion
        (item, google_pay_data): (
            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
            payments::GooglePayWalletData,
        ),
```

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 7]
```diff
@@ -177,12 +199,155 @@ impl From<CardIssuer> for String {
...
+        &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+        payments::Card,
+    )> for BankOfAmericaPaymentsRequest
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        value: (
+            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+            payments::Card,
+        ),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:275]:
```suggestion
        (item, ccard): (
            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
            payments::Card,
        ),
```

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 8]
```diff
@@ -177,12 +199,155 @@ impl From<CardIssuer> for String {
...
+    From<(
+        &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+        Option<String>,
+    )> for ProcessingInformation
+{
+    fn from(
+        value: (
+            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+            Option<String>,
+        ),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:235]:
```suggestion
        (item, payment_solution): (
            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
            Option<String>,
        ),
```

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 9]
```diff
@@ -177,12 +199,155 @@ impl From<CardIssuer> for String {
...
+    From<(
+        &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+        BillTo,
+    )> for OrderInformationWithBill
+{
+    fn from(
+        value: (
+            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
+            BillTo,
+        ),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:212]:
```suggestion
        (item, bill_to): (
            &BankOfAmericaRouterData<&types::PaymentsAuthorizeRouterData>,
            BillTo,
        ),
```

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 5 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 6 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 7 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 8 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 9 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs]: Done

**Review by ArjunKarthik — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by Gnanasundari24 — APPROVED** [review]

**Pull request merged**